Begin implementing Garrison (in feature branch)

RunAwayDSP opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

Co-Authored-by: Zach2good
Additional Coding support from entire ASB dev teams
Additional Capture data from Wiggo
Other Credits: Big thanks to Ophelia, TeoTwawki, and Aerec for help with various aspects along the way
 
## What does this pull request do?
All Zones have a working Garrison

## Included Settings
changing these allows for easier testing as well
    GARRISON_LOCKOUT             = 1800;    -- Time in seconds before a new garrison can be started (default: 1800)
    GARRISON_TIME_LIMIT          = 1800; -- Time in seconds before lose ongoing garrison (default: 1800)
    GARRISON_ONCE_PER_WEEK       = 0;    -- Set to 1 to bypass the limit of one garrison per Conquest Tally Week.
    GARRISON_PARTY_LIMIT         = 18;   -- Set to max party members you want to do garrison (default: 18).
    GARRISON_NATION_BYPASS       = 0;    -- Set to 1 to bypass the nation requirement.
    GARRISON_RANK                = 2;    -- Set to minumum Nation Rank to start Garrison (default: 2).

## What is not included / will not be completed in this PR
Wave Data Captures to get correct wave order
Capture look/name of all NPCs
Update npcs table on Conquest Update with new look and name

## Unable to complete
Check to see if there is Ballista in the Zone
Battlefield Effect (need another way to prevent outside help)
Get names to actually display

## How to play test party size
	--gets party size for spawning each NPC
    party = #party
Change the #party to up to 18

<!-- Clear and detailed steps to test your changes here -->

RunAwayDSP committed:
[WIP] Garrison

Zones that Work*
Ronfaure
Dunes

TODO:
Move win/lose triggers to Guard
Once Per Conquest Tally Limit (plus setting to change)
Check to see if there is Ballista in the Zone
30 minute time limit (trigger loss)
spawn mobs in mini waves also make random from list (not sure how to do this)
need pathing so NPC's return to spawnpoint
All guard event ids (win, loss, wrong nation, tally wait limit)

The-Aerec commented on sql/mob_pools.sql:
```diff
@@ -7089,6 +7089,15 @@ INSERT INTO `mob_pools` VALUES (7036,'Sapphirine_Quadav_LITD','Sapphire_Quadav',
 INSERT INTO `mob_pools` VALUES (7037,'Sapphire_Quadav_LITD','Sapphire_Quadav',202,0x00009F0200000000000000000000000000000000,3,3,5,265,100,0,1,0,1,0,0,0,86,133,0,0,455,0,0,202,202);
 INSERT INTO `mob_pools` VALUES (7038,'Excaliace','Excaliace',145,0x01000503AC10A720A730A740A750006000700000,1,1,5,240,100,0,0,0,0,0,0,32,7,1179,0,0,0,64,256,0,145);
 
+--Garrison Valkune Dunes
```
Needs extra space after the -- in SQL comments

-- Garrison

The-Aerec commented on sql/mob_spawn_points.sql:
```diff
@@ -33122,6 +33122,29 @@ INSERT INTO `mob_spawn_points` VALUES (17187453,'Spriggan_Graverobber','Spriggan
 INSERT INTO `mob_spawn_points` VALUES (17187455,'Spriggan_Graverobber','Spriggan Graverobber',57,0,0,0,0);
 
+
+--Garrison NPCs
```
Here as well

The-Aerec commented on sql/mob_spawn_points.sql:
```diff
@@ -34734,6 +34757,33 @@ INSERT INTO `mob_spawn_points` VALUES (17199659,'Valkurm_Imperator','Valkurm Imp
 INSERT INTO `mob_spawn_points` VALUES (17199661,'Valkurm_Imperator','Valkurm Imperator',73,718.630,-0.500,-83.620,84);
 
+
+--Garrison Traders
```
same as above

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,243 @@
+		npc:getZone():setLocalVar(string.format("[GARRISON]EndTime_%s", zoneID), os.time() + lockout)
+	else 
+		-- TODO event for not having met requirements
+    end
```
Spacing and Styling

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,243 @@
+    local garrisonData = xi.garrison.lookup[zoneId]
+    xi.garrison.waveAlive(player, npc, wave, party)
+end
+
```
remove double spacing between functions

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,243 @@
+			local npcs = garrisonZoneData.npcs
+			local npcnum = 1
+			while npcnum <= party do
+				if GetMobByID(npcs):isAlive()==true then
+					GetMobByID(npcs):addEnmity(GetMobByID(boss), 1, 1)
+					GetMobByID(boss):addEnmity(GetMobByID(npcs), 1, 1)
+				end
+				npcnum = npcnum + 1
+				npcs = npcs + 1
```
spacing

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,243 @@
+    local killedallnpc = true
+    local npcnum = 1
+	local mob = garrisonZoneData.mobs
+    local mobnum = 1
```
Spacing, it seems the alignment is off for most of the script for some reason.

RunAwayDSP committed:
Garrison Spacing

Garrison Spacing...

RunAwayDSP committed:
Garrison Styling

No double spacing between functions

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,235 @@
+        mob = mob + 1
+    end
+    -- If logic im sorry!!!
+    if killedallmobs == true and wave == 4 and killedallnpc == false then
```
See Styling guide on lines with multiple conditions for if/else

The-Aerec replied:
if
    cond1
    cond2
then

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,235 @@
+            xi.garrison.despawnNPCs(npc, party)
+            -- Var to give rewards to party trigger from guard
+            player:setCharVar("Garrison_Treasure", 1)
+    elseif killedallmobs == true and wave <=3 and killedallnpc == false then
```
Same here

The-Aerec commented on scripts/zones/Valkurm_Dunes/npcs/Nyata-Mobuta_WW.lua:
```diff
@@ -15,6 +16,11 @@ local guardEvent  = 32759
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    if(player:getNation()==guardNation) then
```
no need for parenthesis + spacing in conditional

player:getNation() == guardNation

The-Aerec commented on scripts/zones/Valkurm_Dunes/npcs/Quanteilleron_RK.lua:
```diff
@@ -15,6 +16,11 @@ local guardEvent  = 32763
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    if(player:getNation()==guardNation) then
```
Same as above

The-Aerec commented on scripts/zones/Valkurm_Dunes/npcs/Tsunashige_IM.lua:
```diff
@@ -15,6 +16,11 @@ local guardEvent  = 32761
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    if(player:getNation()==guardNation) then
```
same here

The-Aerec commented on scripts/zones/West_Ronfaure/npcs/Doladepaiton_RK.lua:
```diff
@@ -15,6 +16,11 @@ local guardEvent  = 32763
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    if(player:getNation()==guardNation) then
```
same here

The-Aerec commented on scripts/zones/West_Ronfaure/npcs/Kyanta-Pakyanta_WW.lua:
```diff
@@ -15,6 +16,11 @@ local guardEvent  = 32759
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    if(player:getNation()==guardNation) then
```
and here

The-Aerec commented on scripts/zones/West_Ronfaure/npcs/Yoshihiro_IM.lua:
```diff
@@ -15,6 +16,11 @@ local guardEvent  = 32761
 
 entity.onTrade = function(player, npc, trade)
     xi.conq.overseerOnTrade(player, npc, trade, guardNation, guardType)
+    if(player:getNation()==guardNation) then
```
here as well

The-Aerec reviewed (commented):
More styling finds. Getting close! It looks good, hopefully we can expand and finish the implementation

RunAwayDSP committed:
more style issues

more style issues

RunAwayDSP committed:
more spacing

fix spacing

RunAwayDSP commented:
With Battlefield mobs wont aggro (mobs should aggro) so either Battlefield isnt the correct effect or 
you should be able to fight any mob without battlefield and mobs should be able to fight anyone regardless of battlefield 

If it isnt battlefield then we need a new effect that is similar

RunAwayDSP committed:
trailing spaces

RunAwayDSP committed:
Merge pull request #1 from LandSandBoat/base

Grab Upstream

TeoTwawki commented:
> With Battlefield mobs wont aggro (mobs should aggro) so either Battlefield isnt the correct effect or you should be able to fight any mob without battlefield and mobs should be able to fight anyone regardless of battlefield
> 
> If it isnt battlefield then we need a new effect that is similar

I can confirm battlefield is the correct effect. The battlefield ID the system expects to have attached to the effect might not be, but I don't know what that can effect. As long as all entities have matching ID they should be able to interact, and mobs and players that don't cannot - so no outside healing, and no aggro from mobs that aren't part of the garrison. All of that is correct.

On retail these are basically BCNM but without a burning circle zone happening right in the field instead. And more stages than multi part BCNM normally get (typically only 2, and those are mission fights anyway).

RunAwayDSP commented:
> 

I was able to confirm there is no additional effects besides Level_Restriction
There is something else being used to prevent others from helping

TeoTwawki commented:
> > 
> 
> I was able to confirm there is no additional effects besides Level_Restriction There is something else being used to prevent others from helping ![vlcsnap-2022-05-25-11h58m48s870](https://user-images.githubusercontent.com/6879206/170307564-0cd4ea8c-cf89-46bc-bce2-b975e636c49e.png)

:thinking: strange, I had pics form my old group in 2006 with bcnm icon. but the level cap is whats doing it if thats not present - SE does the same thing for seasonal events, must have decided the battlefield effect was redundant. (they ain't wrong, I guess, kinda was..)

Our core does not -yet- have support for using a field in the level cap status for restrict interaction for events, but that needs to change for seasonal battles and now for this as well.

My own server uses check can be added in `HasClaim` in battleutils right before the `bool found = false;` line that would allow any objects with matching values to interact but I do it in a way that is not fit for retail accuracy (I used status effect on the mobs) so I never PR'd it.

RunAwayDSP committed:
Merge pull request #2 from LandSandBoat/base

Update from Upstream

RunAwayDSP committed:
Garrison

Add Settings,
Add Additional Wave Logic,

Code Cleanup
Still a lot to do, im sure there is missing styling

The-Aerec commented:
Please make sure to address CI checks:

Let me know if you need help setting up a local version of the CI lua checks. (It will alleviate alot of comments to run those checks before you upload.)

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+-- Helpers
+-----------------------------------
+
+mobsAlive = function(player)
```
xi.garrison.mobsAlive

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+    end
+end
+
+npcAlive = function(player, party)
```
same as above xi.garrison

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+    end
+end
+
+despawnNPCs = function(npc, party)
```
same as above

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+xi.garrison.tick = function(player, npc, wave, party)
+    print("Garrison Tick wave:", wave)
+    local zoneId = npc:getZoneID()
+    local garrisonData = xi.garrison.lookup[zoneId]
```
This isn't used, is it needed? Perhaps needed for code to be added later?

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+xi.garrison.spawnWave = function(player, npc, wave, party)
+    local zoneId = npc:getZoneID()
+    local garrisonZoneData = xi.garrison.data[zoneId]
+    local mob = garrisonZoneData.mobs
```
Same as above this isn't used. So throwing CI warning

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+    if party > 12 then
+        allianceSize = wave + 2
+    end
+    local mob = garrisonZoneData.mobs
```
repeat call as above thats not used

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+    --TODO spawn mobs in 15 second intervals random from table
+    for _, mobId in ipairs(garrisonZoneData.waveSize[allianceSize]) do
+    npcs = garrisonZoneData.npcs
+    npcnum = 1
```
local npcnum

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+end
+
+xi.garrison.waveAlive = function(player, npc, wave, party)
+    local zoneId = npc:getZoneID()
```
zoneID - You have this used as zoneID throughout the script not ZoneId

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+    local garrisonRunning = npc:getZone():getLocalVar(string.format("[GARRISON]EndTime_%s", zoneID))
+    -- Check all NPCs are dead -> lose
+    if
+        mobsAlive(player) == false and
```
xi.garrson.mobsalive() 

Next conditional also needs similar

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+            player:setCharVar("Garrison_Treasure", 1)
+    elseif
+        mobsAlive(player) == false and
+        wave <=3 and
```
same here as above

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,282 @@
+            mobnum = mobnum + 1
+            mob = mob + 1
+        end
+        despawnNPCs(npc, party)
```
same here

The-Aerec reviewed (commented):
CI checks feedback

RunAwayDSP committed:
Garrison

More Styling
Additional Setting

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,275 @@
+    local npcnum = 1
+    local mob = garrisonZoneData.mobs
+    local mobnum = 1
+    local garrisonRunning = npc:getZone():getLocalVar(string.format("[GARRISON]EndTime_%s", zoneID))
```
zoneID -> zoneId

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,275 @@
+        xi.garrison.npcAlive(player, party) == true
+    then
+        --win
+        npc:getZone():setLocalVar(string.format("[GARRISON]EndTime_%s", zoneID), os.time())
+        npc:getZone():setLocalVar(string.format("[GARRISON]LockOut_%s", zoneID), os.time())
```
Here too

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,275 @@
+        os.time() >= garrisonRunning
+    then
+        -- lose
+        npc:getZone():setLocalVar(string.format("[GARRISON]EndTime_%s", zoneID), os.time())
+        npc:getZone():setLocalVar(string.format("[GARRISON]LockOut_%s", zoneID), os.time())
```
And here

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,275 @@
+    local garrisonZoneData = xi.garrison.data[zoneId]
+    local lockout = xi.settings.GARRISON_LOCKOUT
+    local timeLimit = xi.settings.GARRISON_TIME_LIMIT
+    local garrisonRunning = npc:getZone():getLocalVar(string.format("[GARRISON]LockOut_%s", zoneID))
```
Same

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,275 @@
+        player:confirmTrade()
+        player:setCharVar("GARRISON_CONQUEST_WAIT", getConquestTally())
+        npc:getZone():setLocalVar(string.format("[GARRISON]EndTime_%s", zoneID), os.time() + timeLimit)
+        npc:getZone():setLocalVar(string.format("[GARRISON]LockOut_%s", zoneID), os.time() + lockout)
```
zoneID -> zoneId

The-Aerec commented on scripts/globals/garrison.lua:
```diff
@@ -0,0 +1,275 @@
+            mobnum = mobnum + 1
+            mob = mob + 1
+        end
+        despawnNPCs(npc, party)
```
xi.garrison.

The-Aerec reviewed (commented):
More CI! D:

RunAwayDSP committed:
More Style Issues

More of the same

RunAwayDSP committed:
Garrison

missed one

RunAwayDSP committed:
Garrison

Dynamic??

RunAwayDSP committed:
Garrison

Clean it up!

RunAwayDSP committed:
garrison

shh

RunAwayDSP committed:
Garrison

minor data fixes

RunAwayDSP commented:
> > With Battlefield mobs wont aggro (mobs should aggro) so either Battlefield isnt the correct effect or you should be able to fight any mob without battlefield and mobs should be able to fight anyone regardless of battlefield
> > If it isnt battlefield then we need a new effect that is similar
> 
> ~I can confirm battlefield is the correct effect. The battlefield ID the system expects to have attached to the effect might not be, but I don't know what that can effect.~ _meep! me wuz wrong!_ As long as all entities have matching ID they should be able to interact, and mobs and players that don't cannot - so no outside healing, and no aggro from mobs that aren't part of the garrison. All of that is correct.


https://user-images.githubusercontent.com/6879206/170396620-be0dee5a-f877-4228-b300-e5a89e02fc7a.mp4

shh nobody saw anything

TeoTwawki commented:
was there a point in that? You already showed proof in your previous post that is only used level cap status

TeoTwawki commented:
I did a little late night testing and SE seems to be a little inconstant with themselves (again. shocker I know!)

Most things with sight based detection would link if the mob family was right and would agro if in the right level range to do so. However sound agro'ing mobs weren't. Blood agro also happened. Skeleton in Qufim attacked me. The worst of which was it gave me the restriction i expected saying I could not fight the skeleton - so skelly could hit me, I could not hit skelly. In your video you were able to attack the linked orcs. Likewise I did cape terrigan and would up with a great number of unwanted goblins, but I was able to target them _and_ their pet rarabs.

...I suspect we could do "better than retail" without any complaints, or at least wind up with a very popular module to do so because no server operator is going to want to discover what exactly retail is doing how dumb it is and then explain to their players "no no, that's legit what it does" 
![facepalm](https://user-images.githubusercontent.com/6871475/170539251-07d12071-f4ef-4cee-81a3-fd020c64f692.png)

TeoTwawki commented on scripts/globals/garrison_data.lua:
```diff
@@ -0,0 +1,812 @@
+            },
+            [5] =
+            {
+                17199593, -- Goblin Swordmaker
```
I'm very against having to store all the IDs and other data for each garrison event in one big file like this. There a couple of ways to  skirt around needing the IDs, but there will need to be some discussion before anything is decided; I'll wait for Zach's input especially on this subject since he had a wip for garrison already.

RunAwayDSP replied:
There is no expectation of getting my pushed through was meant to be a draft but i suck at git.

As for the tables of all ids.. Im not a fan either but se's stupid wave logic didnt make it easy... Since its not always first ids that spawn first (fighters in ronfaure are first wave and cursemakers which have lowest IDs dont sure up until they all spawn)

Zachs wave tables kind of work just need a way to increase wave size based on number of alliance

Either way smarter tables are needed

TeoTwawki replied:
gotcha. we'll take what we can of it tho.

RunAwayDSP commented:
I finally figured out how to change it to a draft

RunAwayDSP committed:
Garrison

Ensure same zone for add effects

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
garrison

spaces

RunAwayDSP committed:
Merge branch 'base' of https://github.com/RunAwayDSP/server into base

RunAwayDSP committed:
Garrison

Move triggers to Guard

RunAwayDSP commented:
Latest changes move triggers to guard..
the last TODO: Critical is spawn data and wave data for the west of the zones.

The following are not as critical but make it a little more retail like but does not effect the game play
TODO: Non-Critical Issues
All guard text ids (win, loss, wrong nation, tally wait limit)
Spawn mobs in mini waves also make random from list
Need pathing so NPC's return to spawnpoint
Send NPCs back to spawnpoint if mobs are too far away
Capture look/name of all NPCs
Battlefield Effect (need another way to prevent outside help)
Optimize Code/tables

Essentially at this point it just needs massive play testing to see if there are any critical game play issues

RunAwayDSP committed:
Update Dropid

RunAwayDSP committed:
merge conflict

RunAwayDSP committed:
more conflict updates

RunAwayDSP committed:
Merge branch 'base' of https://github.com/LandSandBoat/server into LandSandBoat-base

RunAwayDSP committed:
garrison update

RunAwayDSP committed:
Merge branch 'LandSandBoat-base' into base

RunAwayDSP committed:
North Gust and West Saru

North Gust wont work until all mobs have spawn pos's

RunAwayDSP committed:
Garrison Cleanup

More Code Cleanup

RunAwayDSP committed:
Garrison Spawn Points

North Gustaberg Spawn Points

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Garrison Remove Effects on Win

RunAwayDSP committed:
Merge branch 'base' of https://github.com/RunAwayDSP/server into base

RunAwayDSP committed:
Garrison Jugner Forrest

Add Jugner Forrest Garrison

RunAwayDSP committed:
Garrison Marshlands

Add Garrison Marshlands

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Garrison Style Cleanup

Cleanup Tables styling

RunAwayDSP committed:
Merge branch 'base' of https://github.com/RunAwayDSP/server into base

RunAwayDSP committed:
Garrison BUBURIMU_PENINSULA

Add BUBURIMU_PENINSULA Garrison

RunAwayDSP committed:
Garrison MERIPHATAUD_MOUNTAINS

Add MERIPHATAUD_MOUNTAINS Garrison

RunAwayDSP committed:
Garrison Qufim

Add Qufim Island Garrison

RunAwayDSP committed:
Garrison OnZone Init Create NPC table

Creates NPCs table on server statup/zoneinit stores all the ID's in garrison_data

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Garrison Cleanup

Remove Unused Vars

RunAwayDSP committed:
Garrison Add BEAUCEDINE_GLACIER

Add BEAUCEDINE_GLACIER
and Move Create NPC Table to zone tick (zone init appears to be too early)

RunAwayDSP committed:
Merge branch 'base' of https://github.com/LandSandBoat/server into base

RunAwayDSP committed:
Garrison THE_SANCTUARY_OF_ZITAH

Add THE_SANCTUARY_OF_ZITAH Garrison

RunAwayDSP committed:
Garrison YUHTUNGA_JUNGLE

Add YUHTUNGA_JUNGLE Garrison

RunAwayDSP committed:
Garrison XARCABARD

Add XARCABARD Garrison

RunAwayDSP committed:
Garrison Add EASTERN_ALTEPA_DESERT

Add EASTERN_ALTEPA_DESERT

RunAwayDSP committed:
Garrison Pathing

Adds Garrison Pathing Thanks to Ophelia

RunAwayDSP committed:
Garrison YHOATOR_JUNGLE

Add YHOATOR_JUNGLE

RunAwayDSP committed:
Garrison Cape Terrigan

Add Cape Terrigan

RunAwayDSP commented:
All zones now work*
Will need heavy play testing (I can only test so many things locally)

I probably wont add anything else to this PR so switching away from draft now

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Garrison Spawn Intervals

Spawn the mobs in intervals during the waves

RunAwayDSP committed:
Garrison NPC Messages and Rank

Add Garrison Messages
Add Rank Requirement

RunAwayDSP committed:
Garrison Fix Bugs

Send NPC back to spawn point
Despawn all npcs before starting incase some are still around.

RunAwayDSP commented:
Cleaned up some bugs I found, Added NPC Dialog.. Added Missing Rank Requirement.. 

Ready for Review/Feedback

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Garrison Add Random Function

Mobs spawn randomly from available table (matches retail behavior now)

RunAwayDSP committed:
Garrison Update NPC Strength

NPCs were too strong this is more accurate

RunAwayDSP committed:
Garrison Get Functioning Names

Issue with renaming is the packet offset for "equipped" not aligning so had to change to a different look in database, and assign a new look on spawn

RunAwayDSP committed:
Garrison Gil Reward

Add Gil Reward for victory

RunAwayDSP committed:
Garrison Gil/Cleanup

Gil text
Code Cleanup/remove unneeded lines
More cleanup to come

RunAwayDSP commented:
> So it seems the only thing holding this back is the Garrison NPC Ally name issue @RunAwayDSP can you post some Screenshots and maybe we can all get together and see if we can figure it out.

Got the name issue figured out has to do with "equipped" look so changed to look 42 (from model_id). Added a few more things did some additional cleanup.

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Merge branch 'LandSandBoat:base' into base

RunAwayDSP committed:
Garrison Text Correction

Fix Outpost Text when starting garrison

TeoTwawki commented:
@The-Aerec 
> So it seems the only thing holding this back is the Garrison NPC Ally name issue @RunAwayDSP can you post some Screenshots and maybe we can all get together and see if we can figure it out.

https://github.com/LandSandBoat/server/pull/1796#pullrequestreview-986537310

the work is probably continuing in the feature branch

zach2good commented:
I've not paid much attention to the dev work that's been going on here, but it's a big piece of work and it is starting to resemble a feature. So I'm going to merge it into `feature/garrison` and clean up the commits and some other bits there.

Sometime after that it'll start getting merged into `everything` for general testing.

zach2good commented:
Ok, I've merged and cleaned up the branch here: https://github.com/LandSandBoat/server/tree/feature/garrison

It's safe to branch that and target it with new PRs

Some initial points:
- Since we've had some issues lately with entire features breaking because they were reliant on IDs not shifting (Full Speed Ahead, my baby :*( ), we're not super-keen on big tables of IDs anymore. The way we're starting to get around this can be seen here: https://github.com/LandSandBoat/server/commit/6da4bc52e94c8366589ad7f460c96333978865e4, where we query by name and then build our various offsets from there. I'm hoping that the mobs that you're storing IDs for have unique names, or at least ONE of them has a unique name so you can use that as the base offset.
- We've stopped using raw item IDs in Lua, we have them listed in `items.lua` now
- Global functions can be more descriptive: `xi.garrison.onRemove` -> `xi.garrison.onStatusRemove`, `xi.garrison.npcTable = function(zone)` -> `xi.garrison.buildNpcTable = function(zone)` etc.
- Commits like "oops", "fixing a typo", etc. aren't particularly helpful to people looking at the history in the future. If you have to add small changes, look into using `git commit --amend` and `git push --force` to your working branch. 

Otherwise, I think there's some good patterns emerging here, good work 👍

zach2good commented:
Also, I think it would be better to comment out all of the other garrison areas, and focus on getting a single one to 100%, rather than continuing to work on each piecemeal.
